### PR TITLE
fix(parser): harden RFC 5321/5322 conformance (CFWS, CR/LF, IP literals)

### LINKED ISSUE #2: test: RFC conformance — isEmail test suite (1226 edge cases)
## Summary

Import the [isEmail](https://github.com/dominicsayers/isemail) XML test suite (1226 edge cases) as our conformance baseline. This is the same suite used by `email-address-parser`.

## Tasks

- [ ] Download `tests/isemail_tests.xml` from isEmail repo
- [ ] Write XML parser → generate Rust test cases (build.rs or codegen script)
- [ ] Map isEmail categories to our `Strictness` levels (ISEMAIL_VALID → Standard, ISEMAIL_RFC5321 → Strict, etc.)
- [ ] Fix any parser bugs found by the suite
- [ ] Target: ≥98% pass rate on ISEMAIL_VALID + ISEMAIL_RFC5321 + ISEMAIL_RFC5322 categories

## Context

Without this, "RFC 5322 compliant" in README is unverified. This is blocking for crates.io publish.

## Time estimate

1d

### PR DESCRIPTION
**Pull request by polaz**
## Summary

Adds the [isEmail](https://github.com/dominicsayers/isemail) conformance suite (v3.05, 164 edge cases) as a baseline and fixes the parser bugs it surfaced. **All 164 cases pass (100%)**; the issue's target categories (VALID + RFC5321 + RFC5322) are 60/60.

This closes both **#2** (conformance suite) and **#40** (leading comment) — the leading-CFWS handling needed for conformance is the same fix as #40, so it is folded in here and PR #44 is superseded.

## Conformance harness

- `tests/isemail_tests.xml` — the suite, embedded via `include_str!`.
- `tests/conformance.rs` — data-driven test mapping each isEmail category onto our `Strictness` ladder (`VALID < DNSWARN < RFC5321 < CFWS < DEPREC < RFC5322 < ERR`). Decodes the U+2400 control-picture glyphs back to real control chars, runs the most-permissive valid config, and gates on **≥98% for VALID + RFC5321 + RFC5322** plus a *zero-unexpected-divergence* check against a documented allowlist (currently empty — full match).

```
VALID 14/14 · DNSWARN 8/8 · RFC5321 16/16 · CFWS 10/10
DEPREC 20/20 · RFC5322 30/30 · ERR 66/66   →   164/164 (100%)
```

## Parser fixes (surfaced by the suite)

- **Leading `[CFWS]` before the local-part** (RFC 5322 `dot-atom = [CFWS] dot-atom-text [CFWS]`) accepted in Standard/Lax; Strict still rejects. *(this is #40)*
- **Drop `input.trim()`** — bare CR/LF are no longer stripped, so `user@host\n` is now **rejected** (header-injection hardening). Valid folding whitespace (`CRLF + WSP`) is still accepted.
- **Address-literal validation** (RFC 5321 §4.1.3): only a valid IPv4 dotted-quad or `IPv6:`-tagged IPv6 (`core::net`) is accepted; general/malformed `[...]` literals are rejected with the new `ErrorKind::InvalidAddressLiteral`.
- **obs-qp / obs-qtext** (control characters) accepted in quoted strings in Lax.
- **Quoted-pair of non-ASCII rejected** (RFC 6531 keeps UTF-8 directly in qtext).
- **Bare CR/LF inside a comment rejected** (only `CRLF + WSP` is valid FWS).

## Behavior changes

- Domain literals must now be valid IP literals (was: any `dtext`).
- Addresses with bare CR/LF are now rejected (was: silently trimmed/accepted).
- `original()` returns the input untrimmed (doc updated).

## Testing

- `cargo nextest run --all-features` → 98 passed
- `cargo nextest run --no-default-features` → 96 passed
- `cargo test --doc --all-features` → 5 passed
- `cargo clippy --all-targets --all-features -- -D warnings` → clean

Closes #2
Closes #40

## Summary by CodeRabbit

* **Bug Fixes**
  * Reject addresses with bare CR/LF and newline sequences that could bypass validation
  * Tightened domain-literal validation: only valid IPv4/IPv6 are accepted; non-IP literals are rejected with a clearer error
  * Improved header-injection hardening by avoiding automatic trimming and tightening comment/CFWS handling
* **Documentation**
  * Updated `original` field documentation to state it is preserved exactly as supplied (not trimmed)
  * Added a **Conformance** section describing validation status against the isEmail v3.05 suite
* **Tests**
  * Added conformance testing against the isEmail v3.05 XML suite and expanded parsing coverage (comments, quoted/display-name behavior, and error expectations)

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -29,6 +29,7 @@ rayon = { version = "1", optional = true }
 [dev-dependencies]
 serde_json = "1"
 criterion = { version = "0.8", features = ["html_reports"] }
+roxmltree = "0.20"
```
[the remark this answers is not in the document]
Comment (polaz) [reply Cargo.toml]: @coderabbitai Done — bumped roxmltree to 0.21.1 (0.21.0 was yanked). Tests and clippy pass.

File: src/parser.rs [hunk 2]
```diff
@@ -143,32 +138,41 @@ pub(crate) fn parse(
...
+    // accepting `user@host\n` would be a header-injection hazard. Legitimate
+    // leading/trailing CFWS (spaces, folding whitespace, comments) is consumed
+    // by skip_cfws in Standard/Lax; a bare CR/LF is left to be rejected.
+    let mut parser = Parser::new(input);
+    let allow_obs = matches!(strictness, Strictness::Lax);
 
     // Try name-addr format: display-name? "<" addr-spec ">"
     let display_name = if allow_display_name {
-        try_parse_display_name(&mut parser)
+        try_parse_display_name(&mut parser, allow_obs)
```
[the remark this answers is not in the document]
Comment (polaz) [reply src/parser.rs:162]: Fixed — leading CFWS is now stripped before display-name/angle detection, so a leading space no longer diverts a quoted display name to the unquoted scanner. Added a regression test (`leading_cfws_before_quoted_display_name`).

File: src/parser.rs [hunk 3]
```diff
@@ -578,43 +578,59 @@ fn parse_domain_label(parser: &mut Parser<'_>) -> Result<(), Error> {
...
 
+/// Returns true if the domain-literal content (the text between `[` and `]`)
+/// is a valid IPv4 address literal or an `IPv6:`-tagged IPv6 address literal
+/// (RFC 5321 §4.1.3). Uses `core::net` parsers (no-std friendly).
+fn is_address_literal(content: &str) -> bool {
+    use core::net::{Ipv4Addr, Ipv6Addr};
+    if let Some(v6) = content.strip_prefix("IPv6:") {
+        return v6.parse::<Ipv6Addr>().is_ok();
+    }
+    content.parse::<Ipv4Addr>().is_ok()
```
[the remark this answers is not in the document]
Comment (polaz) [reply src/parser.rs]: Fixed — the `IPv6:` tag is now matched case-insensitively (eq_ignore_ascii_case), so `[ipv6:::1]` and `[IPV6:...]` parse. Added a regression test (`ipv6_address_literal_tag_is_case_insensitive`).

File: Cargo.toml (hunk 1 above)
[the remark this answers is not in the document]
Comment (polaz) [reply Cargo.toml]: @coderabbitai Done — pinned to exact `roxmltree = "0.21.1"` for reproducible conformance runs.

File: tests/conformance.rs [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+
+    // (pass, total) per category.
+    let mut per_cat: BTreeMap<&str, (u32, u32)> = BTreeMap::new();
+    let mut divergences: Vec<(u32, String, String, bool, Option<String>)> = Vec::new();
+
+    for node in doc.descendants().filter(|n| n.has_tag_name("test")) {
+        let id: u32 = node
+            .attribute("id")
+            .expect("test must have id")
+            .parse()
```
[the remark this answers is not in the document]
Comment (polaz) [reply tests/conformance.rs]: @coderabbitai Switched the three `expect()` calls to `unwrap_or_else(|e| panic!(...))` to match the rest of the test suite. Note: clippy `--all-targets --all-features -- -D warnings` was already clean — the `expect_used` deny is `#![cfg_attr(not(test), deny(...))]` on the lib crate, which does not apply to this integration-test crate. Changed anyway for consistency.
